### PR TITLE
HA, tests, and README

### PR DESCRIPTION
**Pull request by ktsakalozos**
1. Spark HA
2. Scalability and HA amulet tests
3. README fixes

### CODE REVIEW

File: bigtop-packages/src/charm/spark/layer-spark/reactive/spark.py [hunk 1]
```diff
@@ -38,13 +38,15 @@ def report_status():
                            'yarn execution mode not available')
         return
 
-    if mode == 'standalone' and is_state('leadership.is_leader'):
+    if mode == 'standalone' and is_state('zookeeper.ready'):
+        mode = mode + " - HA"
+    elif mode == 'standalone' and is_state('leadership.is_leader'):
```
Comment (pengale) [inline bigtop-packages/src/charm/spark/layer-spark/reactive/spark.py:43]: Is there ever a case where zookeeper.ready is set and leadership.is_leader is also set? If so, is it correct to ignore the leadership.is_leader state, as this code currently does?

**Comment by pengale** [comment]
Running bundletester, I got a linter error that needs to be fixed, as well as some integration tests failures that are probably related to the lxd bug. I'm going to try re-running on AWS. Here's a linky to the error output: http://paste.ubuntu.com/18709482/

File: bigtop-packages/src/charm/spark/layer-spark/tests/10-test-ha.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class TestDeployment(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.d = amulet.Deployment(series='trusty')
+        cls.d.add('sparkscale', 'spark', units=3)
+        cls.d.add('openjdk', 'openjdk')
+        cls.d.add('zk', 'apache-zookeeper', 'zk')
```
Comment (ktsakalozos) [inline bigtop-packages/src/charm/spark/layer-spark/tests/10-test-ha.py]: This is an error you should fix kjackal!

**Comment by pengale** [comment]
@ktsakalozos As you predicted, I still get an error in the scaling test when running with updated code against aws:

```
ERROR: spark::03-scale-standalone.py
[/tmp/bundletester-pzC0Qg/spark/tests/03-scale-standalone.py exit 1]
2016-07-07 12:21:52 Starting deployment of local.aws-test:default
2016-07-07 12:21:53 Deploying services...
2016-07-07 12:21:53  Deploying service sparkscale using /tmp/charm2aoitlrb/trusty/spark
2016-07-07 12:36:19 Adding relations...
2016-07-07 12:36:20  Adding relation openjdk:java <-> sparkscale:java
2016-07-07 12:43:16 Deployment complete in 1283.53 seconds
/home/petevg/.local/lib/python3.5/site-packages/path.py:1717: DeprecationWarning: path is deprecated. Use Path instead.
  warnings.warn(msg, DeprecationWarning)
Timeout occurred, printing juju status...[Services] 
NAME       STATUS EXPOSED CHARM                
openjdk           false   cs:openjdk-4         
spark      active false   local:trusty/spark-0 
sparkscale active false   local:trusty/spark-1 

[Relations] 
SERVICE1    SERVICE2   RELATION   TYPE        
openjdk     spark      java       regular     
openjdk     sparkscale java       regular     
spark       openjdk    java       subordinate 
spark       spark      sparkpeers peer        
sparkscale  openjdk    java       subordinate 
sparkscale  sparkscale sparkpeers peer        

[Units]      
ID           WORKLOAD-STATUS JUJU-STATUS VERSION   MACHINE PORTS              PUBLIC-ADDRESS MESSAGE                     
spark/0      active          idle        2.0-beta8 0       8080/tcp,18080/tcp 107.20.49.103  ready (standalone - master) 
  openjdk/0  active          idle        2.0-beta8                            107.20.49.103  OpenJDK 8 (jre) installed   
sparkscale/0 active          idle        2.0-beta8 1       8080/tcp,18080/tcp 54.224.134.197 ready (standalone)          
  openjdk/3  active          idle        2.0-beta8                            54.224.134.197 OpenJDK 8 (jre) installed   
sparkscale/1 active          idle        2.0-beta8 2       8080/tcp,18080/tcp 54.173.43.246  ready (standalone - master) 
  openjdk/1  active          idle        2.0-beta8                            54.173.43.246  OpenJDK 8 (jre) installed   
sparkscale/2 active          idle        2.0-beta8 3       8080/tcp,18080/tcp 54.84.152.224  ready (standalone)          
  openjdk/2  active          idle        2.0-beta8                            54.84.152.224  OpenJDK 8 (jre) installed   

[Machines] 
ID         STATE   DNS            INS-ID              SERIES AZ         
0          started 107.20.49.103  i-0c3df1e1d6345d76f trusty us-east-1a 
1          started 54.224.134.197 i-018709955e876c267 trusty us-east-1c 
2          started 54.173.43.246  i-0d04956ddab251826 trusty us-east-1a 
3          started 54.84.152.224  i-04bc8f1578b54ddaa trusty us-east-1d 

E
======================================================================
ERROR: test_scaleup (__main__.TestScaleStandalone)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/bundletester-pzC0Qg/spark/tests/03-scale-standalone.py", line 76, in test_scaleup
    "ready (standalone)"]}, timeout=900)
  File "/home/petevg/.local/lib/python3.5/site-packages/amulet/sentry.py", line 598, in wait_for_messages
    for i in helpers.timeout_gen(timeout):
  File "/home/petevg/.local/lib/python3.5/site-packages/amulet/helpers.py", line 108, in timeout_gen
    raise TimeoutError()
amulet.helpers.TimeoutError

----------------------------------------------------------------------
Ran 1 test in 2362.244s

FAILED (errors=1)
Waiting for units to become ready.
Waiting for units to agree on master.
Waiting for the cluster to select a new master.

PASS: 4 ERROR: 1 SKIP:1 Total: 6 (3188.404278 sec)
```

**Comment by ktsakalozos** [comment]
The last commit should remove the flakiness on the 03-scale-standalone test. http://pastebin.ubuntu.com/18782173/

File: bigtop-packages/src/charm/spark/layer-spark/reactive/spark.py (hunk 1 above)
Comment (ktsakalozos) [inline bigtop-packages/src/charm/spark/layer-spark/reactive/spark.py:43]: Yes there is the case where zookeeper ready and is leader is set. This happens when the leaderelection layer has elected the unit to be a leader but at the same time the master spark unit is to be selected via zookeeper. In this state the leader election layer should be ignored as the master will be decided by zookeeper.

**Comment by pengale** [comment]
@ktsakalozos The scaling test works well for me :-) But I'm now getting an error in the final test, when I run on AWS :-(

Error follows:

```
ERROR: spark::10-test-ha.py
[/tmp/bundletester-iW2bw2/spark/tests/10-test-ha.py exit 1]
2016-07-08 14:49:22 Starting deployment of local.aws-test:miles
2016-07-08 14:49:23 Deploying services...
2016-07-08 14:49:23  Deploying service zk using cs:trusty/apache-zookeeper-3
2016-07-08 14:49:32 Adding 1 more units to sparkscale
Traceback (most recent call last):
  File "/usr/local/bin/juju-deployer", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/deployer/cli.py", line 140, in main
    run()
  File "/usr/local/lib/python2.7/dist-packages/deployer/cli.py", line 248, in run
    importer.Importer(env, deployment, options).run()
  File "/usr/local/lib/python2.7/dist-packages/deployer/action/importer.py", line 313, in run
    self.add_units()
  File "/usr/local/lib/python2.7/dist-packages/deployer/action/importer.py", line 67, in add_units
    self.env.add_units(svc.name, abs(delta))
  File "/usr/local/lib/python2.7/dist-packages/deployer/env/go.py", line 52, in add_units
    return self.client.add_units(service_name, num_units)
  File "/usr/local/lib/python2.7/dist-packages/jujuclient.py", line 1356, in add_units
    return self.service.add_units(*args, **kws)
  File "/usr/local/lib/python2.7/dist-packages/jujuclient.py", line 2459, in add_units
    "NumUnits": num_units}})
  File "/usr/local/lib/python2.7/dist-packages/jujuclient.py", line 1482, in rpc
    return self.env._rpc(op)
  File "/usr/local/lib/python2.7/dist-packages/jujuclient.py", line 448, in _rpc
    raise EnvError(result)
jujuclient.EnvError: <Env Error - Details:
 {   u'Error': u'cannot add unit 1/1 to service "sparkscale": cannot add unit to service "sparkscale": service is not alive',
    u'RequestId': 1,
    u'Response': {   }}
 >
/home/petevg/.local/lib/python3.5/site-packages/path.py:1717: DeprecationWarning: path is deprecated. Use Path instead.
  warnings.warn(msg, DeprecationWarning)
E
======================================================================
ERROR: setUpClass (__main__.TestDeployment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/bundletester-iW2bw2/spark/tests/10-test-ha.py", line 33, in setUpClass
    cls.d.setup(timeout=1800)
  File "/home/petevg/.local/lib/python3.5/site-packages/amulet/deployer.py", line 668, in setup
    subprocess.check_call(shlex.split(cmd))
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['juju-deployer', '-W', '-c', '/tmp/amulet-juju-deployer-907rjf8m/deployer-schema.json', '-e', 'local.aws-test:miles', '-t
', '1900', 'local.aws-test:miles']' returned non-zero exit status 1                                                                                              

----------------------------------------------------------------------
Ran 0 tests in 14.901s

FAILED (errors=1)

PASS: 5 ERROR: 1 Total: 6 (1883.747159 sec)
petevg@raven:layer-apache-bigtop-base$ 
```

**Comment by ktsakalozos** [comment]
Thank you for your time you spent in this review.

I see what is happening in your run. Both the scale and HA tests use a spark service called sparkscale. The idea is to kill the master unit in sparkscale and see if a new master is elected. The scale test makes sure it destroyes the sparkscale service (in its teardown method) and the HA deploys the sparkscale service with 3 units again. For some reason in your run the HA test starts before the teardown of the scale test has cleaned up the sparkscale service. There fore is says "I have to add 1 more unit to reach a total of three" but during provisioning the sparkscale service is being wiped out. Testing this theory now ...

**Comment by ktsakalozos** [comment]
Does it seem better now? I 've renamed the service od HA test to sparkha so it will not interfere with the name on the scale test.

**Comment by johnsca** [comment]
While ensuring that the two tests deploy different units is more safe, it does mean the tests will take longer.  I don't understand the "service is not alive" error; I've never seen Juju report that before, and from my (somewhat cursory) reading of the test, it seems like the service should still be up with two units.  Ideally, we'd be able to re-use those and avoid spinning up a whole new set of three machines with the incurred cost of doing so.

**Comment by ktsakalozos** [comment]
We move on with this since the comments were addressed. Merging it as we agreed.

**Pull request merged**